javatests/arcs/core/host/HandleHolderLifecycleTest.kt

```diff
@@ -0,0 +1,255 @@
+      scheduler = scheduler,
+      storageEndpointManager = testStorageEndpointManager(),
+      foreignReferenceChecker = ForeignReferenceCheckerImpl(emptyMap())
+    )
+    val allocator = Allocator.create(hostRegistry, handleManagerImpl, testScope)
+    host.setup()
+
+    allocator
+  }
+
+  fun createHost(
+    schedulerProvider: SchedulerProvider,
+    vararg particles: ParticleRegistration
+  ) = ParticleSpyArcHost(
+    HandleManagerFactory(
+      schedulerProvider,
+      testStorageEndpointManager(),
+      platformTime = FakeTime()
+    ),
+    *particles
+  )
+
+  /** A fake [Particle] that allows one to inject a [HandleHolderSpy] for tests. */
+  @Suppress("UNCHECKED_CAST")
+  class ParticleSpy(val parent: Particle) : Particle {
+    override val handles = HandleHolderSpy(parent.handles as HandleHolderBase)
+  }
+
+  /** A [HandleHolder] with relaxed constraints and method flags used for tests. */
+  class HandleHolderSpy(val parent: HandleHolderBase) : HandleHolderBase("Test", emptyMap()) {
+
+    var detachWasCalled = false
+    var resetWasCalled = false
+
+    init {
+      parent.handles.entries.forEach { (name, handle) -> this.setHandle(name, handle) }
+    }
+
+    override fun getEntitySpecs(handleName: String): Set<EntitySpec<out Entity>> {
+      return parent.getEntitySpecs(handleName)
+    }
+
+    override fun getHandle(handleName: String): Handle {
+      return handles[handleName] ?: handles.values.first()
+    }
+
+    override fun setHandle(handleName: String, handle: Handle) {
+      handles[handleName] = handle
+    }
+
+    override fun detach() {
+      super.detach()
+      detachWasCalled = true
+    }
+
+    override fun reset() {
+      super.reset()
+      resetWasCalled = true
+    }
+  }
+
+  @Test
+  fun startArc_handleHolder_startsOutEmpty() = runBlocking {
+    val host = createHost(schedulerProvider, ::SingleReadHandleParticle.toRegistration())
+    val allocator = setupHost(host)
+
+    var callbackExecuted = false
+    host.registerParticleTransform { particle ->
+      callbackExecuted = true
+      // Assert that the particle starts out empty.
+      assertThat(particle.handles.isEmpty()).isTrue()
+      particle
+    }
+
+    // Calls ArcHost.startArc()
+    allocator.startArcForPlan(SingleReadHandleTestPlan).waitForStart()
+
+    assertThat(callbackExecuted).isTrue()
+  }
+
+  @Test
+  fun startArc_handleHolder_hasAllHandlesSet() = runBlocking {
+    val host = createHost(schedulerProvider, ::MultiHandleParticle.toRegistration())
+    val allocator = setupHost(host)
+
+    var callbackExecuted = false
+    var handleHolder: HandleHolder? = null
+    host.registerParticleTransform { particle ->
+      callbackExecuted = true
+      handleHolder = particle.handles
+      particle
+    }
+
+    // Calls ArcHost.startArc()
+    allocator.startArcForPlan(MultiHandleTestPlan).waitForStart()
+    assertThat(callbackExecuted).isTrue()
+
+    // Handles in handle holder are set
```

Comment:
Question for the reviewer: Is there a good way to compare types on the handles in the holder vs the plan?

Reply:
Unfortunately no. The BaseHandle has a spec object with the type, but this is hidden by the wrappers generated by createHandle(). I suppose you could make a HandleManagerImpl that doesn't wrap in createHandle() and returns the SingletonHandle/CollectionHandle instances directly?

Reply:
I'm satisfied with a no. I was wondering if there was some convention in the codebase. Maybe, we'll figure out a non-kludgy way of testing this later (I imagine this case is covered by unit tests somewhere?).